mathesar/api/db/permissions/schema.py

```diff
@@ -1,21 +1,73 @@
 from django.db.models import Q
 from rest_access_policy import AccessPolicy
 
-from mathesar.models.users import Role
+from mathesar.models.users import DatabaseRole, Role, SchemaRole
 
 
 class SchemaAccessPolicy(AccessPolicy):
+    # Anyone can view a Schema as long as they have
+    # at least a Viewer access to that schema or the database the schema is part of
+    # Create access is restricted to superusers or managers of the database the schema is part of.
+    # This restriction is taken care of the serializer used for creating the Schema
```

Comment:
> Is there a broader picture organizing where given permissions are defined that I'm not understanding, or is the structure driven by implementation details?

[Users & Permissions Spec](https://wiki.mathesar.org/en/product/specs/users-permissions#database-permissions) has a chart of allowed actions based on permissions.

`create` method is an odd one as its access is based not on the object itself rather it is based on an object higher in the hierarchy. Do you have any suggestions on how to make this clear?

Reply:
I don't have a problem with allowed actions per permissions, I'm more concerned with the fact that the places in the codebase where permissions are allowed or denied seem kind of ad hoc. Similar to our approach to general validation, which I know you agree is becoming pretty painful.

Is it possible to reframe `create` under `edit` for Databases?

Reply:
> Is it possible to reframe create for Schemas as edit for Databases, since that's the relevant permission?

Where do you want it to be reframed? `create` is a method in the SchemaViewset and the access policies act on a method of a Viewset, so it is not possible to add this to the Database Access Policy. 

I agree looking at it as `edit` for databases is the correct perspective.

Reply:
> the places in the codebase where permissions are allowed or denied seem kind of ad hoc.

@mathemancer these places are determined by the `drf-access-policy` library we're using (which I chose after evaluating a bunch of different options because it seemed like the simplest for our use case. @silentninja's code seems idiomatic for usage of that library, and I think it makes sense that anyone working on the permissions should familiarize themselves with the library ([the docs](https://rsinger86.github.io/drf-access-policy/) are pretty short).

I do think our comments should make it clear that we're expecting people to be familiar with the library. I haven't actually looked at the code in this PR so I don't know if that's the case or not.